File: flyweight/php/Flyweight.php
<?php

// Flyweight: Character formatting
class CharacterFormat {
    private $font;
    private $size;
    private $color;

    public function __construct(string $font, int $size, string $color) {
        $this->font = $font;
        $this->size = $size;
        $this->color = $color;
    }

    public function display(string $char): string {
        return "Character '{$char}' [Font: {$this->font}, Size: {$this->size}, Color: {$this->color}]";
    }
}

// Flyweight Factory
class FormatFactory {
    private $formats = [];

    public function getFormat(string $font, int $size, string $color): CharacterFormat {
        $key = "{$font}-{$size}-{$color}";
        if (!isset($this->formats[$key])) {
            echo "Created new format: {$key}\n";
            $this->formats[$key] = new CharacterFormat($font, $size, $color);
        }
        return $this->formats[$key];
    }

    public function getTotalFormats(): int {
        return count($this->formats);
    }
}

// Character with extrinsic state
class Character {
    private $char;
    private $format;

    public function __construct(string $char, CharacterFormat $format) {
        $this->char = $char;
        $this->format = $format;
    }

    public function display(): string {
        return $this->format->display($this->char);
    }
}

// Text Editor
class TextEditor {
    private $characters = [];
    private $factory;

    public function __construct() {
        $this->factory = new FormatFactory();
    }

    public function addCharacter(string $char, string $font, int $size, string $color): void {
        $format = $this->factory->getFormat($font, $size, $color);
        $this->characters[] = new Character($char, $format);
    }

    public function display(): array {
        $results = [];
        foreach ($this->characters as $char) {
            $results[] = $char->display();
        }
        $results[] = "Total unique formats: " . $this->factory->getTotalFormats();
        return $results;
    }
}

// Usage
$editor = new TextEditor();
$editor->addCharacter('H', 'Arial', 12, 'Red');
$editor->addCharacter('e', 'Arial', 12, 'Red');
$editor->addCharacter('l', 'Arial', 12, 'Red');
$editor->addCharacter('l', 'Times', 14, 'Blue');
$editor->addCharacter('o', 'Times', 14, 'Blue');

$results = $editor->display();
foreach ($results as $result) {
    echo $result . "\n";
}

File: patterns/flyweight/python/Flyweight.py
from typing import Dict

# Flyweight: Character formatting (intrinsic state)
class CharacterFormat:
    def __init__(self, font: str, size: int, color: str):
        self.font = font
        self.size = size
        self.color = color

    def display(self, char: str) -> str:
        return f"Character '{char}' [Font: {self.font}, Size: {self.size}, Color: {self.color}]"

# Flyweight Factory
class FormatFactory:
    def __init__(self):
        self._formats: Dict[str, CharacterFormat] = {}

    def get_format(self, font: str, size: int, color: str) -> CharacterFormat:
        key = f"{font}-{size}-{color}"
        if key not in self._formats:
            self._formats[key] = CharacterFormat(font, size, color)
            print(f"Created new format: {key}")
        return self._formats[key]

    def get_total_formats(self) -> int:
        return len(self._formats)

# Character with extrinsic state
class Character:
    def __init__(self, char: str, fmt: CharacterFormat):
        self.char = char
        self.format = fmt

    def display(self) -> str:
        return self.format.display(self.char)

# Text Editor
class TextEditor:
    def __init__(self):
        self._characters = []
        self._factory = FormatFactory()

    def add_character(self, char: str, font: str, size: int, color: str):
        fmt = self._factory.get_format(font, size, color)
        self._characters.append(Character(char, fmt))

    def display(self):
        for char in self._characters:
            print(char.display())
        print(f"Total unique formats: {self._factory.get_total_formats()}")


if __name__ == "__main__":
    editor = TextEditor()

    # Adding characters with different formats
    editor.add_character("H", "Arial", 12, "Red")
    editor.add_character("e", "Arial", 12, "Red")
    editor.add_character("l", "Arial", 12, "Red")
    editor.add_character("l", "Times", 14, "Blue")
    editor.add_character("o", "Times", 14, "Blue")

    editor.display()
